keps/sig-network/3836-kube-proxy-improved-ingress-connectivity-reliability/README.md

```diff
@@ -0,0 +1,346 @@
+
+## Summary
+
+The service controller in the Kubernetes cloud controller manager (KCCM)
+configures the load balancer and its corresponding health check (HC) following
+service related events. The configured health check is then used by the load
+balancer as to determine which instances are candidates for traffic load
+balancing. For certain cloud providers (GCP): the KCCM configures the HC to
+target the Kubernetes service proxy for this information. In this KEP we will
+focus on Kube-proxy, since it's the only service proxy under the responsibility
+of the Kubernetes project.
+
+We can define two classes of services for what concerns the HC:
+
+- A) services which have `externalTrafficPolicy: Cluster`
+- B) services which have `externalTrafficPolicy: Local`
+
+For class A) services: Kube-proxy currently returns an answer following its
+`healthz` state. For class B) services: Kube-proxy only reports if the service
+for which the load balancer was created for, has a `Ready` endpoint running on
+the node.
+
+This KEP proposes two changes:
+
+1. That Kube-proxy's includes the `healthz` state AND its current answer w.r.t
+  the local endpoints, when it answers to the HC for class B) services.
+
+2. That Kube-proxy enables connection draining for terminating Nodes. This is to
+  be done by Kube-proxy by inspecting a field on the Node object which indicates
+  that the "node is terminating/deleting" and when seen, starts failing the HC.
+
+```
+<<[UNRESOLVED HC-linked-service-fields ]>>
+We need to agree on this field. `deletionTimestamp` is the perfect indicator for
+this, but....the Node object does not have a finalizer, so that wouldn't help.
+@thockin advocates for `.spec.unschedulable`, but I am inclined to say "that
+doesn't indicate if a Node is going away". `unschedulable` just says that no
+workloads can be assigned to the Node anymore. It is true that it's usually
+followed up by a drain...but I am not sure it's good to implement something
+which can easily kill ingress connectivity to a Node / cluster by simply doing
+`kubectl cordon node ...` (It would kill ingress connectivity to the entire
+cluster if the ingress controller is running on the node which is being cordoned)
+
+Maybe we need a KEP for a finalizer on the Node object (*troll face)
+<<[/UNRESOLVED]>>
+```
+
+```
+<<[UNRESOLVED HC-linked-service-fields ]>>
+Figure out which other cloud-providers configure the HC to target the service
+proxy for the HC
+<<[/UNRESOLVED]>>
+```
+
+## Motivation
+
+The motivation for each change is:
+
+1. Not including the `healthz` state can cause Kube-proxy to indicate to a load
+  balancer that it should send traffic to a Node simply because the endpoint is
+  scheduled there, even though Kube-proxy might not be healthy and successfully
+  managed to write the rules required for actually being able to forward traffic
+  to the endpoint.
+
+2. Workloads on terminating Nodes would benefit from connection draining as to
```

Comment:
This is less about workloads ON those nodes and more about those nodes serving LB traffic in general.

Example:

Node A has 1 backend pod for service X.
Node B has other stuff but not X pods.
Service  X is `eTP: Cluster`

Client 1 connects to External LB.  LB chooses node B.  Node B chooses pod on node A.

Node B is cordoned.  HC starts failing.

Node B draining begins.

Client 1's connection continues.

Client 2 connects to External LB.  LB chooses node A because B is HC-failing.

Client 1 finishes.

Node B drainign finishes.

Node B is deleted.

The pods on node B were irrelevant, but the fact that it was an intermediate hop for other traffic was important.  Without the HC failing, LBs could send new connections to node B right up until the node was deleted, and then they all get broken.

Reply:
Good point, I overlooked this